helix-core/src/history.rs

```diff
@@ -22,72 +58,195 @@ impl Default for History {
-                children: SmallVec::new(),
-                revert: Transaction::from(ChangeSet::new(&Rope::new())),
+                last_child: 0,
+                transaction: Transaction::from(ChangeSet::new(&Rope::new())),
+                inversion: Transaction::from(ChangeSet::new(&Rope::new())),
+                timestamp: Instant::now(),
             }],
-            cursor: 0,
+            current: 0,
         }
     }
 }
 
 impl History {
     pub fn commit_revision(&mut self, transaction: &Transaction, original: &State) {
         // TODO: could store a single transaction, if deletes also stored the text they delete
-        let revert = transaction
+        let inversion = transaction
             .invert(&original.doc)
             // Store the current cursor position
             .with_selection(original.selection.clone());
 
-        let new_cursor = self.revisions.len();
+        let new_current = self.revisions.len();
+        self.revisions[self.current].last_child = new_current;
         self.revisions.push(Revision {
-            parent: self.cursor,
-            children: SmallVec::new(),
-            revert,
+            parent: self.current,
+            last_child: 0,
+            transaction: transaction.clone(),
+            inversion,
+            timestamp: Instant::now(),
         });
-
-        // add a reference to the parent
-        self.revisions
-            .get_mut(self.cursor)
-            .unwrap() // TODO: get_unchecked_mut
-            .children
-            .push((new_cursor, transaction.clone()));
-
-        self.cursor = new_cursor;
+        self.current = new_current;
     }
 
     #[inline]
     pub fn current_revision(&self) -> usize {
-        self.cursor
+        self.current
     }
 
     #[inline]
     pub const fn at_root(&self) -> bool {
-        self.cursor == 0
+        self.current == 0
     }
 
     pub fn undo(&mut self) -> Option<&Transaction> {
         if self.at_root() {
-            // We're at the root of undo, nothing to do.
             return None;
         }
 
-        let current_revision = &self.revisions[self.cursor];
+        let current_revision = &self.revisions[self.current];
+        self.current = current_revision.parent;
+        Some(&current_revision.inversion)
+    }
 
-        self.cursor = current_revision.parent;
+    pub fn redo(&mut self) -> Option<&Transaction> {
+        let current_revision = &self.revisions[self.current];
+        let last_child = current_revision.last_child;
+        if last_child == 0 {
+            return None;
+        }
+        self.current = last_child;
 
-        Some(&current_revision.revert)
+        let last_child_revision = &self.revisions[last_child];
+        Some(&self.revisions[last_child].transaction)
     }
 
-    pub fn redo(&mut self) -> Option<&Transaction> {
-        let current_revision = &self.revisions[self.cursor];
+    fn lowest_common_ancestor(&self, mut a: usize, mut b: usize) -> usize {
```

Comment:
Should we use euler tour of tree here? I wonder if it's worth it.

Reply:
Do you mean to leverage the parallel algorithm for finding the lowest common ancestor? I don't see a benefit. We don't need to optimize the running time here. Even if we wanted to, the algorithm running time is also bounded by the path length - we need to output all of the transactions up and down so optimizing the LCA doesn't improve the overall asymptotic time.